### PR TITLE
ChemMaster and CondiMaster refactored, prints items over time

### LINKED ISSUE #59734: Cherry wine cannot be interacted with in a chemmaster
Reporting client version: 513.1526

## Round ID:
[164373](https://scrubby.melonmesa.com/round/164373)

## Testmerges:
- [Refactors dbcore and limits the maximum amount of concurrent async queries to a variable amount](https://www.github.com/tgstation/tgstation/pull/59676)

## Reproduction:
Make some cherry wine.
Insert it into a chemmaster.
Try to move it into a beaker and watch as nothing happens.
Seems runtimes are popping up for the majority of the chemicals too when using the chemmaster (or sapmaster if you're a botanist).
Spawned in fruitwine does not appear to have this issue, though.

### PR DESCRIPTION
**Pull request by MTandi**
ZceBfkJaFG

## About The Pull Request

Chem master has been using some hard coded styles for containers, pills and patches. Some people didn't bother to update styles, and they stayed outdated.

Now it uses just a list of reagent containers instead, which makes it easy to add new containers - just one line of code.

Also changed its sprite for better animations, emissive overlay, and to display buffer reagents on the sprite itself.

Now it plays printing animation when you actually print containers and servos have a purpose now - they increase the amount of containers printed at once by up to 100% * part_tier. Pills and patches are always printed at double amount.

And fixes #59734 

## Why It's Good For The Game

Makes chem master easier to manage in code and easier to use in-game.

Printing delay is added for consistency with other fabrication machines (like lathes), to give a purpose for the servo motor used in construction, for visuals (to show the printing animation only when something is being printed, and not just when you move stuff in and out of buffer as before) and to prevent people from spamming the world with unneeded objects with 0 second delay.

## Changelog

:cl:
qol: Chemmaster UI tweaked
fix: Fixed chem master showing wrong data during reagent analysis
fix: Fixed chem master not working with fermented drinks
image: Chemmaster resprited, now has an indicator for buffer reagents
refactor: Refactored chemmaster code, it now uses reagent containers instead of styles
balance: Chemmaster now uses servos and has printing animation of 0.75 second duration. Outputs 1 container or up to 2 pills/patches per cycle by default. Can be upgraded to output up to 4 and 8 with t4 servos.
/:cl:

### CODE REVIEW

**Comment by Wallemations** [comment]
image

Putting this here for posterity in relation to the sprite

**Comment by newfren** [comment]
taking time to mass print pills is really lame!

**Comment by AtomTheProphet** [comment]
> taking time to mass print pills is really lame!

Honestly, if they're going to do this then better parts should increase the quantity of pills, patches, etc., printed at a time.

**Comment by MTandi** [comment]
> > taking time to mass print pills is really lame!
> 
> A bit, yeah. Honestly, if they're going to do this then better parts should increase the quantity of pills, patches, etc., made each time it prints.

Made servo reduce it down to a fraction of a second.

**Comment by ExcessiveUseOfCobblestone** [comment]
I'm gonna have to sleep on the delay

**Comment by Jolly-66** [comment]
This is just a minor nitpick but could you sort out the "dispensed" items in the UI into categories? Having them all lumped up like that looks a little overwhelming

**Comment by MTandi** [comment]
> This is just a minor nitpick but could you sort out the "dispensed" items in the UI into categories? Having them all lumped up like that looks a little overwhelming

I removed them on purpose because I thought that there is no point in showing categories and it saves space. Why do you think categories are needed here?

**Comment by YehnBeep** [comment]
Even with a short delay, mass printing say a vaccine...

**Review by Wallemations — APPROVED** [review]
Sprites look good

**Comment by Jolly-66** [comment]
> @ MTandi

It's just a personal nitpick, that being, I like it when things are organized/presented in an organized fashion. By the icons being organized it just looks nicer/more pleasing, at least to me.

I don't know if other people share this but that's just my 2 cents on the entire matter.

**Comment by MTandi** [comment]
> Even with a short delay, mass printing say a vaccine...

You will need it once or twice in a shift. And it's quite fast, just not instant. I understand that any second taken away from players will cause dissatisfaction, but this makes sense that chem master doesn't dispense containers, but creates them, and it should take a bit of time like with protolathes. More reason for stock parts to exist.

**Comment by YehnBeep** [comment]
> > Even with a short delay, mass printing say a vaccine...
> 
> You will need it once or twice in a shift. And it's quite fast, just not instant. I understand that any second taken away from players will cause dissatisfaction, but this makes sense that chem master doesn't dispense containers, but creates them, and it should take a bit of time like with protolathes. More reason for stock parts to exist.

Honestly it's not so much the raw time that concerns me so much as say... having to sit there and push print 5 times across 75 seconds, the runs long enough to make someone twiddle their thumbs but too short to leave it and go do something else. And, ok, with full upgrades that's only about 20 seconds but - again less about the time cost - it's a marked increase in tedium.

**Comment by MTandi** [comment]
> It's just a personal nitpick, that being, I like it when things are organized/presented in an organized fashion. By the icons being organized it just looks nicer/more pleasing, at least to me.

Here, made this just for you ❤️ 

image

**Comment by MTandi** [comment]
> So, maybe a solution: increase the number of pills that can be printed at once to say 50 or 80 or so, so it can be 'set it and forget it' while it prints?

Hm, OK, I'll make it possible to print up to 50 pills/patches, and containers will be limited to 10 per request

**Comment by MTandi** [comment]
> > So, maybe a solution: increase the number of pills that can be printed at once to say 50 or 80 or so, so it can be 'set it and forget it' while it prints?
> 
> Hm, OK, I'll make it possible to print up to 50 items

Now it kinda implies that there should be "cancel" button...

**Comment by YehnBeep** [comment]
> Now it kinda implies that there should be a "cancel" button...

Might cause some drama when the chemist hits cancel on the virologist's healing virus or vaccine. Or when the clown convinces the hop to give them chem access or steals a medical ID and starts cancelling things. Now, whether that drama's a good or bad thing... I'll leave that to others to discuss/decide.

**Comment by MTandi** [comment]
> Might cause some drama when the chemist hits cancel on the virologist's healing virus or vaccine. Or when the clown convinces the hop to give them chem access or steals a medical ID and starts cancelling things. Now, whether that drama's a good or bad thing... I'll leave that to others to discuss/decide.

It will rather cause frustration when people schedule something by mistake and will deconstruct chem master instead of waiting.

Works fine.

**Comment by ArcaneMusic** [comment]
Any idea if this fixes either of these?
#59734 #69282

**Comment by MTandi** [comment]
> Any idea if this fixes either of these? #59734 #69282

It does not, but I know how to fix first.

Right now chem master works with reagents by mapping them to datums, instead of using actual refs to reagents. So custom reagents don't work.

I will need to display the purity of reagent later so that chefs can tell synthetic milk from cow  milk, and it will be fixed automatically when I add purity lookup to chemmaster analysis.

**Comment by Helg2** [comment]
Maybe also make the package number decrease when objects get printed so you can start the process again, without adjusting the needed number?

**Comment by MTandi** [comment]
> Maybe also make the package number decrease when objects get printed so you can start the process again, without adjusting the needed number?

Tricky to do because of number being stored on FE. The assumption is that when someone stops printing, they have reasons not to resume.

**Comment by ExcessiveUseOfCobblestone** [comment]
can you give me a reason why the delay was added and why its good for the game? Im a bit in a bad spot because i dont want to argue over it on here when youve also added sprite changes and code improvements.

15 seconds for 10 1u pills is a bit rough.

**Comment by MTandi** [comment]
Fixed #59734 and added purity lookup for future cooking changes.

pcWjYYRlsO

**Comment by epochayur** [comment]
I really dislike that such a huge balance change to the ChemMasters is being snuck in with a UI change.
Printing pills/patches is a huge part of doing Medbay right. Anyone who has played doctor or chemist has had scenarios where every single second counts trying to get a cure or a vaccine out the door.
The only reason you've given seems to be:
>More reason for stock parts to exist.
This just seems like tedium for tedium's sake, and I really fail to see how it improves the game or addresses any existing balance issue. 

This change should be added in a separate PR.

**Comment by epochayur** [comment]
> > I really dislike that such a huge balance change to the ChemMasters is being snuck in with a UI change. Printing pills/patches is a huge part of doing Medbay right. Anyone who has played doctor or chemist has had scenarios where every single second counts trying to get a cure or a vaccine out the door. The only reason you've given seems to be:
> > > More reason for stock parts to exist.
> > 
> > 
> > This just seems like tedium for tedium's sake, and I really fail to see how it improves the game or addresses any existing balance issue.
> > This change should be added in a separate PR.
> 
> Don't make it sound so dramatic. You can print pills for an entire station in a matter of seconds. I'm sure you will manage to survive.

Thank you for the extremely dismissive response. You still haven't addressed why you want to time gate this specific mechanic, and why you're so against moving the balance portion of this PR out to a separate PR.

**Comment by KM-Catman** [comment]
Why is a UI Change being combined with a fundamental change to chemistry? 

Also, if you want to slow down production of literally everything from chemistry, why are you targeting the ChemMaster?

**Comment by KM-Catman** [comment]
> > Thank you for the extremely dismissive response. You still haven't addressed why you want to time gate this specific mechanic, and why you're so against moving the balance portion of this PR out to a separate PR.
> 
> I did address it in comments above, please read.
> 
> 1. It has a printing animation that shows how extruder creates a container,
> 2. other machines that fabricate objects already have a delay - see protolathes.
> 3. ChemMaster circuit also requires servo, but it didn't affect the functionality in any way. Now it does.
> 4. I already made a few tweaks according to the feedback to negate the impact on vaccine use case
> 
> I didn't create a separate PR because this change is too minor for a separate PR and is tied to sprites and code that I changed.
> 
> If 10 seconds will lead to a station wipeout or you will be just extremely frustrated to go back and pick up printed pills, unwrench it and let it print outside of medbay. Be creative.

It isn't a "Minor" change though, It's a production cap on items normally created in extreme bulk. You have also yet to mention why such a change is necessary outside of "Other Machines have delays", Notably bringing up the protolathe which typically is used to create items in a FAR smaller quantity than that of the ChemMaster

**Comment by ExcessiveUseOfCobblestone** [comment]
Was purity an existing display? I always used the spec machine.

**Comment by YehnBeep** [comment]
> Was purity an existing display? I always used the spec machine.

It is not, and HPLC machines are very nearly useless and will be pretty much effectively useless with purity displayed here.

But, personally, I would take that as more that HPLCs could use a rework to not be utterly useless outside of something you can print a portable analyzer for anyways...

**Comment by ExcessiveUseOfCobblestone** [comment]
I would say purity of stuff when it can be a "different" reagent has a p powerful use. niche for sure but i am not a big fan of it being on the master hence why I asked

**Comment by ExcessiveUseOfCobblestone** [comment]
Also for future reference please do not put in p big stuff like adding delays to things under the guise of a sprite/refactor PR. While I understand what youre trying to convey, I disagree with the sentiment that it had to be included in this PR.

my current thought is im ok with keeping the delay in so I can get a feel for it both med and service but i will go ahead and say if it gets merged that aspect is definitely not set in stone.

**Review by ExcessiveUseOfCobblestone — DISMISSED** [review]
purity needs to go, otherwise this is ok (delay is subject to change post-merge).

**Comment by MTandi** [comment]
> purity needs to go, otherwise this is ok (delay is subject to change post-merge).

Purity is for https://hackmd.io/@ZSLHD7nTTlWNegwCVVdJBw/ryqRgDluj

**Comment by ExcessiveUseOfCobblestone** [comment]
You can add a warning instead where that matters, the chem master is already overtuned with functionality, it doesnt need to expose purity too when we have machines for that.

**Comment by MTandi** [comment]
Changed purity display from exact percentage to scale: 0-50% - bad, 50-75% average, 70%+ - good.

image

**Comment by MTandi** [comment]
> You can add a warning instead where that matters, the chem master is already overtuned with functionality, it doesnt need to expose purity too when we have machines for that.
> 
> Was this doc given a greenlight by a maintainer (am I stepping on toes)?

I don't want to add a chromatography machine to the kitchen just to let chef determine cow milk from Biogen milk. Do we need a greenlight to work on something? I never got anyone's approval for any of my features and they seem to work fine.

**Comment by ExcessiveUseOfCobblestone** [comment]
> I don't want to add a chromatography machine to the kitchen just to let chef determine cow milk from Biogen milk. Do we need a greenlight to work on something? I never got anyone's approval for any of my features and they seem to work fine.

who said you had to? the purity will be constant 0.5 presumably (unless they mix with real ingredients), they dont need to know their exact purity unless im missing something.

It was more of a am I pushing back on something someone else said was ok, which it doesnt sound like to be the case.

**Comment by MTandi** [comment]
> who said you had to? the purity will be constant 0.5 presumably (unless they mix with real ingredients), they dont need to know their exact purity unless im missing something.

It will for milk, but grown food has variable purity depending on endurance and lifespan of seed.

> Sorry by warning i meant feedback that crafted milk or whatever wont be pure, not to retain purity readings with vague indicators.

I consider adding prefixes like "Synthetic milk", "Milk", "Pure milk", then I may remove these readings from Chem Master. Is it okay to leave them for now as above? 

I don't engage with fermichem a lot so I have no clue why showing purity here is an issue.

**Comment by ExcessiveUseOfCobblestone** [comment]
sure assuming it isnt exposing inverted chems or similar.

**Comment by MTandi** [comment]
Changed the printing speed to printing amount for better visuals, total speed unchanged.

Animation is always 1.5 sec long, printing 1 container or up to 2 pills/patches per cycle. Can be upgraded to up to 4 containers or 8 pills/patches per cycle.

**Comment by Learner10** [comment]
> > > Thank you for the extremely dismissive response. You still haven't addressed why you want to time gate this specific mechanic, and why you're so against moving the balance portion of this PR out to a separate PR.
> > 
> > 
> > I did address it in comments above, please read.
> > 
> > 1. It has a printing animation that shows how extruder creates a container,
> > 2. other machines that fabricate objects already have a delay - see protolathes.
> > 3. ChemMaster circuit also requires servo, but it didn't affect the functionality in any way. Now it does.
> > 4. I already made a few tweaks according to the feedback to negate the impact on vaccine use case
> > 
> > I didn't create a separate PR because this change is too minor for a separate PR and is tied to sprites and code that I changed.
> > If 10 seconds will lead to a station wipeout or you will be just extremely frustrated to go back and pick up printed pills, unwrench it and let it print outside of medbay. Be creative.
> 
> It isn't a "Minor" change though, It's a production cap on items normally created in extreme bulk. You have also yet to mention why such a change is necessary outside of "Other Machines have delays", Notably bringing up the protolathe which typically is used to create items in a FAR smaller quantity than that of the ChemMaster

Yes, because clearly the endless piles of vaccine pills constantly sitting outside medical isn't indicative of a reliance of 'bag and dump' levels of thought. From a 10 unit vial of vaccine, you can see 100 pills made from it... sitting on the floor. For a station of 50 crew that like maybe 20 of them might eat.

Past vaccines, you're rarely going to use the pharmacy for anything that isn't going to be like past 5 pills of iron. Why? Because plumbing exists, and most chemists use plumbing for most of the general mainstay medicines, which by the way, also has a delay on printing pills from the plumbing pill factory.

**Comment by YehnBeep** [comment]
> Because plumbing exists, and most chemists use plumbing for most of the general mainstay medicines, which by the way, also has a delay on printing pills from the plumbing pill factory.

Can't say this matches my experience. From what I've seen... The vast majority of chemists seem to prefer making medicines conventionally since it's faster and gets you higher purity even with minimal effort towards purity.

**Comment by KM-Catman** [comment]
> > > > Thank you for the extremely dismissive response. You still haven't addressed why you want to time gate this specific mechanic, and why you're so against moving the balance portion of this PR out to a separate PR.
> > > 
> > > 
> > > I did address it in comments above, please read.
> > > 
> > > 1. It has a printing animation that shows how extruder creates a container,
> > > 2. other machines that fabricate objects already have a delay - see protolathes.
> > > 3. ChemMaster circuit also requires servo, but it didn't affect the functionality in any way. Now it does.
> > > 4. I already made a few tweaks according to the feedback to negate the impact on vaccine use case
> > > 
> > > I didn't create a separate PR because this change is too minor for a separate PR and is tied to sprites and code that I changed.
> > > If 10 seconds will lead to a station wipeout or you will be just extremely frustrated to go back and pick up printed pills, unwrench it and let it print outside of medbay. Be creative.
> > 
> > 
> > It isn't a "Minor" change though, It's a production cap on items normally created in extreme bulk. You have also yet to mention why such a change is necessary outside of "Other Machines have delays", Notably bringing up the protolathe which typically is used to create items in a FAR smaller quantity than that of the ChemMaster
> 
> Yes, because clearly the endless piles of vaccine pills constantly sitting outside medical isn't indicative of a reliance of 'bag and dump' levels of thought. From a 10 unit vial of vaccine, you can see 100 pills made from it... sitting on the floor. For a station of 50 crew that like maybe 20 of them might eat.
> 
> Past vaccines, you're rarely going to use the pharmacy for anything that isn't going to be like past 5 pills of iron. Why? Because plumbing exists, and most chemists use plumbing for most of the general mainstay medicines, which by the way, also has a delay on printing pills from the plumbing pill factory.

If there ain't a single damn reason for the change, then the change doesn't need to be a thing. No one is complaining about a lack of delay. No one is begging for this change. It doesn't have any reason to exist. No one has brought up a single actual reason that warrants this random choice in an otherwise decent **UI CHANGE**.

**Comment by MTandi** [comment]
> If there ain't a single damn reason for the change, then the change doesn't need to be a thing. No one is complaining about a lack of delay. No one is begging for this change. It doesn't have any reason to exist. No one has brought up a single actual reason that warrants this random choice in an otherwise decent **UI CHANGE**.

Can you please read the discussion? I stated multiple reasons for this change, yet on the other hand I never heard anything from you guys why would this change be bad for the game.

It seems like the only reason you're angry is that I took away a few seconds of your time that you can easily find how to spend, like yell in the chat "I'm making vaccine" or something else. I gave you a workaround for your use case - unwrench and drag the ChemMaster outside and let it print, if you REALLY need to make pill above needed amount and don't want to attend ChemMaster. These pills are barely used, you always see that they're just thrown onto the ground and this pile stays until the end of shift.

**Comment by zoomachina** [comment]
> > If there ain't a single damn reason for the change, then the change doesn't need to be a thing. No one is complaining about a lack of delay. No one is begging for this change. It doesn't have any reason to exist. No one has brought up a single actual reason that warrants this random choice in an otherwise decent **UI CHANGE**.
> 
> Can you please read the discussion? I stated multiple reasons for this change, yet on the other hand I never heard anything from you guys why would this change be bad for the game. Why would I care if someone is complaining about lack of delay? I don't make PRs by community requests or for your amusement. I believe that this change is warranted by game design and it is beneficial for the game. Any tiny nerf to anything will make someone unhappy, that's not the reason to not make the change. State reasons why this may be bad for QOL or make things illogical, or screw up certain use case, and I'll tweak the solution if it's warranted.
> 
> It seems like the only reason you're angry is that I took away a few seconds of your time that you can easily find how to spend, like yell in the chat "I'm making vaccine" or something else. I gave you a workaround for your use case - unwrench and drag the ChemMaster outside and let it print, if you REALLY need to make pill above needed amount and don't want to attend ChemMaster. These pills are barely used, you always see that they're just thrown onto the ground and this pile stays until the end of shift.

Oh yeah? Let me answer, then, even though people already did:
> It has a printing animation that shows how extruder creates a container,
"I made the animation, so we **must** use it!"
> other machines that fabricate objects already have a delay - see protolathes.
Lathes make tools that don't expire once used. Chemmasters are used to make consumables, and yes, sometimes in bulk. You still haven't said why you want to nerf them like that.

> ChemMaster circuit also requires servo, but it didn't affect the functionality in any way. Now it does.
That doesn't matter, though didn't even add that initially.
> I already made a few tweaks according to the feedback to negate the impact on vaccine use case
Do you think vaccines and splitting chemicals are the only use-cases of chemmasters, and they are a decoration otherwise?

**Comment by MTandi** [comment]
>"I made the animation, so we **must** use it!"
It had this animation before. It just was misused

> Lathes make tools that don't expire once used. Chemmasters are used to make consumables, and yes, sometimes in bulk. 
Not all containers are consumed, and you print pills/patches only once - to contain your reagent. This is not an argument.

>You still haven't said why you want to nerf them like that.
I don't consider this as a nerf that affects anyone, but as a consistency fix.

> > ChemMaster circuit also requires servo, but it didn't affect the functionality in any way. Now it does.
> 
> That doesn't matter, though you didn't even add that initially.

The machine starts using more power once upgraded. So there was HARM in upgrading chem master before my change. Now you benefit from upgrading it, so it does matter. 

> > I already made a few tweaks according to the feedback to negate the impact on vaccine use case
> 
> Do you think vaccines and splitting chemicals are the only use-cases of chemmasters, and they are a decoration otherwise?

Vaccines is the only use case where this delay actually affects anything. I made adjustments according to feedback to accelerate that use case. If you want a medicine pill factory - do plumbing.

> This is yet another change that makes life a little more painful for the players, for no reason other than that a coder said so. We shouldn't have to give a reason why this change is bad, you should give a reason why it's good instead, and the fact there's a sprite for it isn't enough.

I gave my reasons, other than sprite. You didn't give any reasons why it's bad other than just saying that you dislike it and mentioning some pain, which is in-constructive and does not require adjustments. If a couple seconds of waiting bring pain to your life, you should consider delegating vaccine creation duty to someone else.

And again as @ExcessiveUseOfCobblestone said, you can always tweak it later. Make a PR to revert this change or reduce the delay if the pain will be that unbearable.

**Comment by zoomachina** [comment]
> It had this animation before. It just was misused

Still an invalid reason to add it in


> Not all containers are consumed, and you print pills/patches only once - to contain your reagent. This is not an argument.

That's dumb as fuck. Not only do people not "print patches and pills only once", reusing test tubes is clunky at best, because they are named and because filling each one of them is even slower than using chemmaster with your change. 

> I don't consider this as a nerf that affects anyone, but as a consistency fix.

Chemmasters not being consistent with lathes is apparently a problem now? Lathes make larger and more complex objects than chemmasters, the only thing that unites them is that they make something, but if that's something that should take time for "consistency", you might as well add a cooldown to the chem dispenser, or, better yet, make chemmaster consume iron or glass. I'm sure it's a much-needed and much-awaited change to the game!

> The machine starts using more power once upgraded. So there was HARM in upgrading chem master before my change. Now you benefit from upgrading it, so it does matter.

Yeah, so harmful it's not noticeable in gameplay. Instead, we should add something very noticeable and make upgrading make it less severe, but not fully, because..?

> Vaccines is the only use case where this delay actually affects anything. I made adjustments according to feedback to accelerate that use case. If you want a medicine pill factory - do plumbing.

Plumbing is not something many people understand or want to understand, nor is it a complete alternative since free space in chemistry is limited and making factories is much more time-consuming. Also, tell me, if you're someone who allegedly wants to "prevent people from spamming the world with unneeded objects"), wouldn't plumbing only worsen that "problem"?

> I gave my reasons, other than sprite. You didn't give any reasons why it's bad other than just saying that you dislike it and mentioning some pain, which is in-constructive and does not require adjustments. If a couple seconds of waiting bring pain to your life, you should consider delegating vaccine creation duty to someone else.

Nobody likes waiting for something, even less so waiting because a coder said so. I can only assume you don't care about that if you really don't see why people dislike this "fix".

> And again as @ExcessiveUseOfCobblestone said, you can always tweak it later. Make a PR to revert this change or reduce the delay if the pain will be that unbearable.

Or you could just not add a bad change into the game from the get-go, rather than wait until it's obvious it's bad, then wait more until a revert pr is not immediately closed because "it's not enough time to tell for sure it's bad!", then wait more until what shouldn't have been added in the first place gets removed.

**Comment by YehnBeep** [comment]
> Also, tell me, if you're someone who allegedly wants to "prevent people from spamming the world with unneeded objects"), wouldn't plumbing only worsen that "problem"?

Chem presses have an internal storage so they don't spam objects out there. I'm not sure what the max is but I've seen it go up to over 800 pills stored.

**Comment by KM-Catman** [comment]
Zoomachina, thank you a lot for summing up pretty much everything wrong with this PR. love ya.

>> Also, tell me, if you're someone who allegedly wants to "prevent people from spamming the world with unneeded objects"), wouldn't plumbing only worsen that "problem"?

> Chem presses have an internal storage so they don't spam objects out there. I'm not sure what the max is but I've seen it go up to over 800 pills stored.

But also, Chem Presses actually dump their entire contents on to the floor when destroyed. I've done it with up to like 200-300 patches that had been contaminated without any issues of lag.

**Comment by YehnBeep** [comment]
> But also, Chem Presses actually dump their entire contents on to the floor when destroyed. I've done it with up to like 200-300 patches that had been contaminated without any issues of lag.

Oh that's pretty interesting, didn't know that. Thanks for sharing the info.

**Comment by MTandi** [comment]
So much talk and still no actual example of why would it be an issue. Okay you need to wait quite a bit if your aim is to print 200-300 pills/patches. For what? Who is going to use them?

You need to wait until your fried egg cooks. Fried egg is one-use, y'know? You need to wait for thermomachine to change that gas temperature, you need materials and waiting to have your shit printed in lathe, why chemists are special and can't wait a second for containers printed for free?

Let's try this then.

@KM-Catman @zoomachina what are your suggestions?

1. Block this PR entirely?
2. Remove the delay entirely, removing animation and leaving servo be because it is unnoticable that it's harmful to upgrade it?
3. Make some other adjustments to delay or pill/patch printing on top of what I already did (pills/patches output is doubled and you can improve speed by 400% with part upgrades)
4. Something else?

**Comment by KM-Catman** [comment]
Just remove the delay part! Even cobblestone said there's no reason to combine it with the UI change, it's the one thing everyone has said to do.

File: code/modules/reagents/chemistry/machinery/chem_master.dm
Comment (ExcessiveUseOfCobblestone) [inline code/modules/reagents/chemistry/machinery/chem_master.dm:177]:
```suggestion
	if(default_deconstruction_crowbar(item))
```
You return so no need for else

File: code/modules/reagents/chemistry/machinery/chem_master.dm
Comment (ExcessiveUseOfCobblestone) [inline code/modules/reagents/chemistry/machinery/chem_master.dm:179]:
```suggestion
	if(is_reagent_container(item) && !(item.item_flags & ABSTRACT) && item.is_open_container())
```
same here all if statements above early return

File: code/modules/reagents/chemistry/machinery/chem_master.dm
Comment (ExcessiveUseOfCobblestone) [inline code/modules/reagents/chemistry/machinery/chem_master.dm:207]: 2 birds 1 stone?

```DM
if(!user?.transferItemToLoc(new_beaker, src))
```

File: code/modules/reagents/chemistry/machinery/chem_master.dm
Comment (ExcessiveUseOfCobblestone) [inline code/modules/reagents/chemistry/machinery/chem_master.dm:392]: p sure QDELETED(src) has !src built in

```suggestion
	if(!item_name || !reagents.total_volume || QDELETED(src) || !usr.can_perform_action(src, ALLOW_SILICON_REACH))
```

**Comment by MTandi** [comment]
> Just remove the delay part! Even cobblestone said there's no reason to combine it with the UI change, it's the one thing everyone has said to do.

There is a reason for the delay not being in a separate PR. If there is no delay, then there is no need for printing animation. If there is no printing animation then there is no need in that 3D printer part in the machine sprite. It makes me wish to change it to something like vending machine, since instant container creation assumes that the containers are pre-prepared and the machine is just stocked with them. I won't do this, I like its current look.

I reduced the delay down to 0.75 sec default, 0.1875 upgraded. So it takes 4.5 seconds to print 50 pills when upgraded. Please deal with it for now.

**Comment by KM-Catman** [comment]
> > Just remove the delay part! Even cobblestone said there's no reason to combine it with the UI change, it's the one thing everyone has said to do.
> 
> There is a reason for the delay not being in a separate PR. If there is no delay, then there is no need for printing animation. If there is no printing animation then there is no need in that 3D printer part in the machine sprite. It makes me wish to change it to something like vending machine, since instant container creation assumes that the containers are pre-prepared and the machine is just stocked with them. I won't do this, I like its current look.
> 
> I halved the delay down to 0.75 sec default, 0.1875 upgraded. So it takes 4.5 seconds to print 50 pills when upgraded. Please deal with it for now.
> 
> This is how it looks like: chemmaster

alright, then if option 2  apparently doesn't work. Go for option 1.

**Comment by MTandi** [comment]
> alright, then if option 2 apparently doesn't work. Go for option 1.

No

**Comment by carshalash** [comment]
Bottles dispensed by the hoochmaster are not displaying the name of the reagent in the bottle, and just auto suggesting to naming it 'bottle'.  This is kind of a drag for tending.

**Pull request merged**
